packages/flutter_tools/lib/src/dart/source_graph.dart

```diff
@@ -0,0 +1,257 @@
+        LibraryMap dartLibraryMap = reader.readFromFile(
+          new JavaFile(librariesFile.path), librariesFile.readAsStringSync()
+        );
+
+        String libPath = path.normalize(path.absolute(path.join(dartSdkPath, 'lib')));
+        for (SdkLibrary library in dartLibraryMap.sdkLibraries)
+          _dartLibraryMap[library.shortName] = path.join(libPath, library.path);
+      }
+    }
+
+    // Parse the .packages file.
+    _packagesFile = new PackagesFile(new File(path.join(directory.path, '.packages')));
+    _packagesFile.parse();
+
+    GraphSource root = _getCreateSource(path.join(directory.path, entryPointPath));
+    root.parse();
+
+    // Parse any dirty sources.
+    while (sources.any((GraphSource source) => source.dirty)) {
+      for (GraphSource source in sources.where((GraphSource source) => source.dirty).toList()) {
+        source.parse();
+      }
+    }
+  }
+
+  GraphSource _getCreateSource(String resolvedPath) {
+    if (_sourceMap.containsKey(resolvedPath)) {
+      return _sourceMap[resolvedPath];
+    } else {
+      GraphSource source = new GraphSource(this, resolvedPath);
+      _sourceMap[resolvedPath] = source;
+      return source;
+    }
+  }
+
+  String _resolveDirectiveUri(String uriString, String basePath) {
+    if (!uriString.contains(':')) {
+      return path.normalize(path.join(path.dirname(basePath), uriString));
+    } else {
+      Uri uri = Uri.parse(uriString);
+
+      if (uri.scheme == 'dart') {
+        if (parseEmbedderSource) {
+          // dart:
+          if (_packagesFile.canResolveDart) {
+            return _packagesFile.resolveDart(uri.toString());
+          } else {
+            return _dartLibraryMap[uri.toString()];
+          }
+        } else {
+          return null;
+        }
+      } else if (uri.scheme == 'package') {
+        // package:
+        String root = uri.pathSegments.first;
+        String relPath = uri.pathSegments.sublist(1).join('/');
+        return _packagesFile.resolvePackage(root, relPath);
+      } else {
+        return null;
+      }
+    }
+  }
+}
+
+class GraphSource {
+  GraphSource(this.graph, this.fullpath);
+
+  final SourceGraph graph;
+  final String fullpath;
+  final List<GraphSourceReference> references = <GraphSourceReference>[];
+  bool dirty = true;
+  DateTime _stamp;
+  String _cachedSource;
+
+  bool get isUpToDate => _file.lastModifiedSync() == _stamp;
+
+  String getSource() {
+    if (_cachedSource != null && isUpToDate)
+      return _cachedSource;
+
+    _cachedSource = _file.readAsStringSync();
+    _stamp = _file.lastModifiedSync();
+    return _cachedSource;
+  }
+
+  void parse() {
+    dirty = false;
+    references.clear();
+
+    CompilationUnit unit = parseDirectives(getSource(), name: fullpath, suppressErrors: true);
+
+    for (Directive directive in unit.directives) {
+      if (directive is ExportDirective || directive is ImportDirective || directive is PartDirective) {
```

Comment:
If you write `if (directive is UriBasedDirective) {`, then you don't need a local variable or the implicit cast.